chisel/src/config.rs

```diff
@@ -0,0 +1,553 @@
+    modules: VecDeque<(String, ModuleConfig)>,
+}
+
+/// A key-value module configuration.
+pub struct ModuleConfig(HashMap<String, String>);
+
+impl ChiselConfig {
+    pub fn rulesets(&self) -> &VecDeque<(String, Ruleset)> {
+        &self.0
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut VecDeque<(String, Ruleset)> {
+        &mut self.0
+    }
+}
+
+impl Ruleset {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.opts
+    }
+
+    pub fn options_mut(&mut self) -> &mut HashMap<String, String> {
+        &mut self.opts
+    }
+
+    pub fn modules(&self) -> &VecDeque<(String, ModuleConfig)> {
+        &self.modules
+    }
+
+    pub fn modules_mut(&mut self) -> &mut VecDeque<(String, ModuleConfig)> {
+        &mut self.modules
+    }
+}
+
+impl ModuleConfig {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.0
+    }
+}
+
+impl FromArgs for ChiselConfig {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let mut ret = VecDeque::new();
+        match Ruleset::from_args(modules, options) {
+            Ok(rs) => {
+                ret.push_back(("cli".to_string(), rs));
+                Ok(ChiselConfig(ret))
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl FromArgs for Ruleset {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        // Filter out whitespace for names.
+        let modules: String = modules.chars().filter(|c| *c != ' ').collect();
+
+        // Collect module names.
+        let module_names: Vec<String> = modules
+            .split(',')
+            .filter_map(|c| match c {
+                "" => None, // If the user put a comma somewhere weird, we need to filter out the empty string.
+                any => Some(any.to_string()),
+            })
+            .collect();
+
+        // Filter out whitespace for options.
+        let options: String = options.chars().filter(|c| *c != ' ').collect();
+
+        // Collect module options.
+        let module_options: Vec<String> = options
+            .split(',')
+            .filter_map(|c| match c {
+                "" => None,
+                any => Some(any.to_string()),
+            })
+            .collect();
+
+        let mut module_configs = VecDeque::new();
+        // Iterate over each module and generate a config from its provided options.
+        for name in module_names.iter() {
+            // Iterate over the collected options and find the options corresponding to this
+            // module.
+            let relevant_options: Vec<String> = module_options
+                .iter()
+                .filter_map(|option| {
+                    // Split by dot here to isolate the module name.
+                    let option_pair: Vec<&str> = option.split('.').collect();
+                    assert!(option_pair.len() >= 1);
```

Comment:
It's `len = 1` if there was no dot, right?

Reply:
Yes, this assertion is there just to crash instead of doing something unpredictable with a zero-length vec. It should not be possible to reach, but in case something changes in the future I put it there.